hcl/fmt.go

```diff
@@ -0,0 +1,153 @@
+	p := hclparse.NewParser()
+	_, diags := p.ParseHCL([]byte(src), filename)
+	if err := errors.L(diags).AsError(); err != nil {
+		return "", errors.E(ErrHCLSyntax, err)
+	}
+	// For now we just use plain hclwrite.Format
+	// but we plan on customizing formatting in the near future.
+	return string(hclwrite.Format([]byte(src))), nil
+}
+
+// FormatTree will format all Terramate configuration files
+// in the given tree starting at the given dir. It will recursively
+// navigate on sub directories. Directories starting with "." are ignored.
+//
+// Only Terramate configuration files will be formatted.
+//
+// Files that are already formatted are ignored. If all files are formatted
+// this function returns an empty result.
+//
+// All files will be left untouched. To save the formatted result on disk you
+// can use FormatResult.Save for each FormatResult.
+func FormatTree(dir string) ([]FormatResult, error) {
+	logger := log.With().
+		Str("action", "hcl.FormatTree()").
+		Str("dir", dir).
+		Logger()
+
+	logger.Trace().Msg("listing terramate files")
+
+	files, err := listTerramateFiles(dir)
+	if err != nil {
+		return nil, errors.E(errFormatTree, err)
+	}
+
+	results := []FormatResult{}
+	errs := errors.L()
+
+	for _, f := range files {
+		logger := log.With().
+			Str("file", f).
+			Logger()
+
+		logger.Trace().Msg("reading file")
+
+		path := filepath.Join(dir, f)
+		fileContents, err := os.ReadFile(path)
+		if err != nil {
+			errs.Append(err)
+			continue
+		}
+
+		logger.Trace().Msg("formatting file")
+
+		currentCode := string(fileContents)
+		formatted, err := Format(currentCode, path)
+		if err != nil {
+			errs.Append(err)
+			continue
+		}
+
+		if currentCode == formatted {
+			logger.Trace().Msg("file already formatted")
+			continue
+		}
+
+		logger.Trace().Msg("file needs formatting, adding to results")
+
+		results = append(results, FormatResult{
+			path:      path,
+			formatted: formatted,
+		})
+	}
+
+	dirs, err := listTerramateDirs(dir)
+	if err != nil {
+		errs.Append(err)
+		return nil, errors.E(errFormatTree, errs)
+	}
+
+	for _, d := range dirs {
+		logger := log.With().
+			Str("subdir", d).
+			Logger()
+
+		logger.Trace().Msg("recursively formatting")
+		subres, err := FormatTree(filepath.Join(dir, d))
+		if err != nil {
+			errs.Append(err)
+			continue
+		}
+		results = append(results, subres...)
+	}
+
+	if err := errs.AsError(); err != nil {
+		return nil, err
+	}
+	return results, nil
+}
+
+// Save will save the formatted result on the original file, replacing
+// its original contents.
+func (f FormatResult) Save() error {
+	return os.WriteFile(f.path, []byte(f.formatted), 0644)
+}
+
+// Path is the absolute path of the original file.
+func (f FormatResult) Path() string {
+	return f.path
+}
+
+// Formatted is the contents of the original file after formatting.
+func (f FormatResult) Formatted() string {
+	return f.formatted
+}
+
+const (
+	errFormatTree errors.Kind = "formatting tree"
```

Comment:
we were placing this at the top in all other packages. Just curious why in the bottom in this one.

Reply:
By default I never placed any private symbols/constants at the top =P. I'm just used to first listing all public things and then later all private things. But can also put on the top for consistency (not a strong opinion).

Reply:
We are adding on the top all exported error sentinels/kinds etc. But this is just my default mode of organizing things, very personal, but I do make a distinction between a private vs public error sentinel, the private one can't even be used by caller for any checking, it is just to avoid duplication internally.

Reply:
ah ok, all the others are public indeed.